### PR TITLE
ifm-web: init at 4.0.2

### PR DESCRIPTION
**Pull request by litchipi**
## Description of changes

Context: https://github.com/NixOS/nixpkgs/pull/328205#pullrequestreview-2236375650

Create a proper package for `ifm`, with the name `ifm-web` as ifm was already taken.
Allows for more meta informations, as well as licensing

Also creates a serve script, that allowed to greatly improve the module serving script.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/if/ifm-web/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+in
+stdenv.mkDerivation {
+  pname = "ifm";
+  inherit version src;
+
+  phases = "installPhase";
+  installPhase = ''
+    mkdir -p $out/bin $out/php
+    ln -s $src $out/php/index.php
```
Comment (SuperSandro2000) [inline pkgs/by-name/if/ifm-web/default.nix]: We should copy the files to not retain a source reference and allow changes to the files.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -5436,7 +5436,7 @@ with pkgs;
 
   ifm = callPackage ../tools/graphics/ifm { };
 
-  ifm-web = callPackage ../by-name/if/ifm-web/package.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:5439]: This line is not required and is done by nixpkgs code already.

**Review by SuperSandro2000 — DISMISSED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

File: pkgs/by-name/if/ifm-web/default.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    SERVE_DIR=$(dirname "$0")/../php/
+    IFM_ROOT_DIR="$3" ${lib.getExe php83} -S "$1:$2" -t "$SERVE_DIR"
+  '';
+
+in
+stdenv.mkDerivation {
+  pname = "ifm";
+  inherit version src;
+
+  phases = "installPhase";
```
[the remark this answers is not in the document]
Comment (litchipi) [reply pkgs/by-name/if/ifm-web/default.nix]: I'm affraid that's something I do regularly on my own nix derivations, didn't know it was a problem

**Comment by litchipi** [comment]
Fixed and squashed to a clean commit

@SuperSandro2000 Is the naming of my commit alright now ? I tried to follow the contributing guidelines.

That would be cool to have a CI for this tho, not the first time I mess up the commit messages

File: pkgs/by-name/if/ifm-web/package.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+in
+stdenv.mkDerivation {
+  pname = "ifm";
+  inherit version src;
+
+  installPhase = ''
+    mkdir -p $out/bin $out/php
+    cp $src $out/php/index.php
+    cp ${serve_script} $out/bin/ifm
+  '';
```
Comment (pbsds) [inline pkgs/by-name/if/ifm-web/package.nix:33]:
```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin $out/php
    cp $src $out/php/index.php
    cp ${serve_script} $out/bin/ifm
    runHook postInstall
  '';
```

File: nixos/modules/services/web-apps/ifm.nix [hunk 5]
```diff
@@ -55,24 +53,20 @@ in {
...
-        ${lib.getExe php} -S ${cfg.listenAddress}:${builtins.toString cfg.port} -t /tmp/ifm
-      '';
-
       serviceConfig = {
         DynamicUser = true;
         User = "ifm";
         StandardOutput = "journal";
         BindPaths = "${cfg.dataDir}:/data";
         PrivateTmp = true;
+        ExecStart = "${lib.getExe pkgs.ifm-web} ${cfg.listenAddress} ${builtins.toString cfg.port} /data";
```
Comment (pbsds) [inline nixos/modules/services/web-apps/ifm.nix:69]:
```suggestion
        ExecStart = "${lib.getExe pkgs.ifm-web} ${lib.escapeShellArg cfg.listenAddress} ${builtins.toString cfg.port} /data";
```

File: pkgs/by-name/if/ifm-web/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/by-name/if/ifm-web/default.nix]: It disables things like patches or fixupPhase

File: pkgs/by-name/if/ifm-web/package.nix [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin $out/php
+    cp $src $out/php/index.php
+    cp ${serve_script} $out/bin/ifm
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Comment (gador) [inline pkgs/by-name/if/ifm-web/package.nix:37]:
```suggestion
  meta = {
```

File: pkgs/by-name/if/ifm-web/package.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    description = "Improved File Manager, a single-file web-based filemanager";
+    longDescription = ''
+      The IFM is a web-based filemanager, which comes as a single file solution using HTML5, CSS3, JavaScript and PHP.
+    '';
+    homepage = "https://github.com/misterunknown/ifm";
+    changelog = "https://github.com/misterunknown/ifm/releases/tag/v${version}";
+    license = licenses.mit;
```
Comment (gador) [inline pkgs/by-name/if/ifm-web/package.nix:44]:
```suggestion
    license = lib.licenses.mit;
```

File: pkgs/by-name/if/ifm-web/package.nix [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    maintainers = with maintainers; [ litchipi ];
```
Comment (gador) [inline pkgs/by-name/if/ifm-web/package.nix:45]:
```suggestion
    maintainers = with lib.maintainers; [ litchipi ];
```

**Review by gador — DISMISSED** [review]
Small nitpicks, otherwise looks good!

File: nixos/modules/services/web-apps/ifm.nix [hunk 9]
```diff
@@ -61,18 +59,13 @@ in {
         IFM_ROOT_DIR = "/data";
```
Comment (pbsds) [inline nixos/modules/services/web-apps/ifm.nix:59]:
```suggestion
```

File: pkgs/by-name/if/ifm-web/package.nix [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  dontUnpack = true;
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin $out/php
+    cp $src $out/php/index.php
+    cp ${serve_script} $out/bin/ifm
+    runHook postInstall
+  '';
+
```
Comment (pbsds) [inline pkgs/by-name/if/ifm-web/package.nix:38]:
```suggestion

  passthru.tests = {
    inherit (nixosTests) ifm;
  };

```

File: pkgs/by-name/if/ifm-web/package.nix [hunk 11]
```diff
@@ -0,0 +1,50 @@
+{
+  fetchurl,
+  stdenv,
+  lib,
+  php83,
+  writeShellScript,
```
Comment (pbsds) [inline pkgs/by-name/if/ifm-web/package.nix:6]:
```suggestion
  writeShellScript,
  nixosTests,
```

**Comment by pbsds** [comment]
in reference to https://github.com/NixOS/nixpkgs/pull/334559#pullrequestreview-2237574618, i believe `ifm-web: init at 4.0.2` is appropriate

**Comment by litchipi** [comment]
Fixed, squashed and rebased.

I also allowed myself to fix the release notes (the "Available as services.ifm" was missing)

**Comment by gador** [comment]
failing test for the manual:
```
bad local reference, id #opt-services.ifm not known
```

**Comment by pbsds** [comment]
thank you!

**Pull request merged**